### PR TITLE
Changed implementation TaskLocal#bind in terms of Task#bracket

### PR DESCRIPTION
**Pull request by leandrob13**
Addresses #601 

Changes the `TaskLocal#bind ` implementation in terms of `Task#bracket` which handles any cleanup of resources upon success, failure or cancellation.

```scala
def bind[R](value: A)(task: Task[R]): Task[R] =
    Task.suspend {
      val saved = ref.value
      ref.update(value)
      task.bracket(r => Task.now(r))(_ => restore(saved))
    }
```

### CODE REVIEW

File: monix-eval/shared/src/main/scala/monix/eval/TaskLocal.scala [hunk 1]
```diff
@@ -134,7 +134,7 @@ final class TaskLocal[A] private (default: => A) {
     Task.suspend {
       val saved = ref.value
       ref.update(value)
-      task.doOnFinish(_ => restore(saved))
+      task.bracket(r => Task.now(r))(_ => restore(saved))
```
Comment (oleg-py) [inline monix-eval/shared/src/main/scala/monix/eval/TaskLocal.scala]: If you want to handle cancellation, `task` has to be used inside bracket. E.g.

```scala
Task.unit.bracket(_ => task)(_ => restore(saved))
```

Also `bindL` needs to have similar change, so maybe it's easier to implement `bind` as `bindL(Task.now(value))`

**Comment by leandrob13** [comment]
@oleg-py  So I redefined bindL like this:

```scala
def bindL[R](value: Task[A])(task: Task[R]): Task[R] = {
    val saved = ref.value
    value.bracket { v =>
      ref.update(v)
      task
    }(_ => restore(saved))
  }
```

Also, took care of `bindClear` too.

File: monix-eval/shared/src/main/scala/monix/eval/TaskLocal.scala [hunk 2]
```diff
@@ -170,12 +166,13 @@ final class TaskLocal[A] private (default: => A) {
...
     *        reset to the previous value
     */
-  def bindL[R](value: Task[A])(task: Task[R]): Task[R] =
-    value.flatMap { value =>
-      val saved = ref.value
-      ref.update(value)
-      task.doOnFinish(_ => restore(saved))
-    }
+  def bindL[R](value: Task[A])(task: Task[R]): Task[R] = {
+    val saved = ref.value
```
Comment (alexandru) [inline monix-eval/shared/src/main/scala/monix/eval/TaskLocal.scala]: This `ref.value` should be suspended as well ...

```scala
Task.eval(ref.value).flatMap { saved =>
  value.bracket {
    ref.update(v)
    task
  } { _ =>
    restore(saved)
  }
}
```

File: monix-eval/shared/src/main/scala/monix/eval/TaskLocal.scala [hunk 3]
```diff
@@ -140,11 +140,7 @@ final class TaskLocal[A] private (default: => A) {
     *        reset to the previous value
     */
   def bind[R](value: A)(task: Task[R]): Task[R] =
-    Task.suspend {
-      val saved = ref.value
-      ref.update(value)
-      task.doOnFinish(_ => restore(saved))
-    }
+    Task.suspend(bindL(Task.now(value))(task))
```
Comment (alexandru) [inline monix-eval/shared/src/main/scala/monix/eval/TaskLocal.scala]: If you suspend that value as I mentioned below, you don't need to suspend here.

Comment (leandrob13) [reply monix-eval/shared/src/main/scala/monix/eval/TaskLocal.scala]: @alexandru  Understood, will take care of it later today.

**Comment by leandrob13** [comment]
@alexandru should I trigger another build?

**Comment by alexandru** [comment]
Please merge with master.

**Comment by alexandru** [comment]
Not 100% sure, but I think this is fine. Will merge for now.

**Pull request merged**